Userland/Utilities/tac.cpp

```diff
@@ -0,0 +1,95 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/Vector.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/File.h>
+#include <unistd.h>
+
+void print_lines(const Vector<String>& vec);
+
+int main(int argc, char** argv)
+{
+    if (pledge("stdio rpath", nullptr) < 0) {
+        perror("pledge");
+        return 1;
+    }
+
+    Vector<const char*> paths;
+
+    Core::ArgsParser args_parser;
+    args_parser.set_general_help("Concatenate files or pipes to stdout, last line first.");
+    args_parser.add_positional_argument(paths, "File path", "path", Core::ArgsParser::Required::No);
+    args_parser.parse(argc, argv);
+
+    Vector<String> lines;
+    if (paths.is_empty()) {
+        char buf[32768];
+        ssize_t nread = read(0, buf, sizeof(buf));
+        if (nread == 0)
+            return 1;
+        if (nread < 0) {
+            perror("read");
+            return 2;
+        }
+
+        auto vec = String(buf, Chomp).split('\n', true);
+        for (auto&& s : vec) {
+            lines.append(move(s));
+        }
+    }
+
+    if (!lines.is_empty()) {
+        print_lines(lines);
+    }
+
+    if (!paths.is_empty()) {
+        for (const char* path : paths) {
+            auto file = Core::File::construct(path);
+            if (!file->open(Core::IODevice::ReadOnly)) {
+                warnln("Failed to open {}: {}", path, file->error_string());
+                return false;
+            }
+
+            Vector<String> lines;
+            while (file->can_read_line()) {
+                auto line = file->read_line();
+                lines.append(line);
+            }
+
+            print_lines(lines);
+        }
+    }
+
+    return 0;
+}
+
+void print_lines(const Vector<String>& vec)
```

Comment:
instead of extracting this for loop into a seperate function, you can just unconditionally do it at the end of the main function, and remove both the calls